support-frontend/assets/components/subscriptionCheckouts/addressSearch/addressComponentStore.js

```diff
@@ -0,0 +1,262 @@
+// @flow
+
+// ----- Imports ----- //
+import { combineReducers, type Dispatch } from 'redux';
+
+import { fromString, type IsoCountry } from 'helpers/internationalisation/country';
+import { type SetCountryAction, setCountry } from 'helpers/page/commonActions';
+import {
+  formError,
+  type FormError,
+  nonEmptyString,
+  notNull,
+  removeError,
+  validate,
+} from 'helpers/subscriptionsForms/validation';
+import { type RegularPaymentRequestAddress } from 'helpers/paymentIntegrations/readerRevenueApis';
+import { type Scoped } from 'helpers/scoped';
+
+import { type AddressType } from 'helpers/subscriptionsForms/addressType';
+import {
+  postcodeFinderReducerFor,
+  type PostcodeFinderState,
+} from 'components/subscriptionCheckouts/address/postcodeFinderStore';
+import type { Option } from 'helpers/types/option';
+import { setFormSubmissionDependentValue } from 'helpers/subscriptionsForms/checkoutFormIsSubmittableActions';
+import { postcodeIsWithinDeliveryArea } from 'helpers/deliveryCheck';
+import type { FulfilmentOptions } from 'helpers/productPrice/fulfilmentOptions';
+
+// ----- Types ----- //
+
+export type FormFields = {|
+  ...RegularPaymentRequestAddress
+|};
+export type FormField = $Keys<FormFields>;
+export type FormErrors = FormError<FormField>[];
+
+type AddressFieldsState = {|
+  ...FormFields,
+  formErrors: FormErrors,
+|}
+
+export type State = {|
+  fields: AddressFieldsState,
+  postcode: PostcodeFinderState
+|};
+
+export type SetCountryChangedAction = { type: 'SET_COUNTRY_CHANGED', country: IsoCountry, ...Scoped<AddressType> };
+
+export type Action =
+  | { type: 'SET_ADDRESS_LINE_1', lineOne: string, ...Scoped<AddressType> }
+  | { type: 'SET_ADDRESS_LINE_2', lineTwo: string, ...Scoped<AddressType> }
+  | { type: 'SET_TOWN_CITY', city: string, ...Scoped<AddressType> }
+  | { type: 'SET_STATE', state: string, ...Scoped<AddressType> }
+  | SetCountryChangedAction
+  | { type: 'SET_ADDRESS_FORM_ERRORS', errors: FormError<FormField>[], ...Scoped<AddressType> }
+  | { type: 'SET_POSTCODE', postCode: string, ...Scoped<AddressType> };
+
+
+// ----- Selectors ----- //
+
+const getPostcodeForm = (state: State): PostcodeFinderState => state.postcode;
+const getStateFormErrors = (state: State): FormErrors => state.fields.formErrors;
+
+const getFormFields = (state: State): FormFields => ({
+  lineOne: state.fields.lineOne,
+  lineTwo: state.fields.lineTwo,
+  city: state.fields.city,
+  country: state.fields.country,
+  postCode: state.fields.postCode,
+  state: state.fields.state,
+});
+
+
+// ----- Functions ----- //
+
+const isPostcodeOptional = (country: Option<IsoCountry>): boolean =>
+  country !== 'GB' && country !== 'AU' && country !== 'US' && country !== 'CA';
```

Comment:
minor: `['GB', 'AU', 'US', 'CA'].contains(country)` if that kind of function exists in JS?

Reply:
Would need to be `!(['GB', 'AU', 'US', 'CA'].includes(country))` not sure if that's any better, the `!` at the start is quite easy to miss

Reply:
yeah I guess the whole function is negative logic - be tempting to rename to `isPostcodeRequired`, then only negate it when we pass into the component right at the end.  But I think we're in to diminishing returns and it's definitely not a blocker!